### DISCUSSION TITLE
I am trying to do this - converting from wstring works incorrectly!

### DISCUSSION THREAD

**Discussion by lipasite** [nlohmann/json#2643] [Q&A] [answered]
I am trying to do this - converting from wstring works incorrectly!
I need add to json array of objects **wstring** values,
and then convert to plain **wchar_t** string!

But all I try - didnt work! This is [screenshot](https://imgur.com/a/mzHiCJI)! 

```
auto jsonProducts = json::array();
while (...) {
  std::wstring productRID_w (ValByName(GoodsQ));
  std::wstring productName_w (ValByName2(GoodsQ));
  // convert to json
  json product = {
	{"productRID", productRID },
	{"productName", productName}
  };
jsonProducts.push_back(product);
}
std::string jStr = jsonProducts.dump();
std::wstring widestr = std::wstring(jStr.begin(), jStr.end());
wchar_t* tempWide = const_cast<wchar_t*>(widestr.c_str());
return tempWide;
```

So, the wstring **L"test"** after adding to jon object becomes **"Р\xfҐP\a"** 

What the hell, man?
How to fix it?
Or suggest a normal C ++ library that can work correctly with wstring.
Be so kind....

_Originally posted by @lipasite in https://github.com/nlohmann/json/discussions/2641_

**Comment by nlohmann** [comment, marked as answer]
This library does not support `std::wstring`, but only `std::string` with UTF-8 encoding.

**Reply by lipasite** [reply]
So, I found the solve:

```
auto jsonProducts = json::array();
while (...) {
			std::wstring productID (wchar_t getProductID(...));
                        std::wstring productID (wchar_t getProductName(...));
			// convert to json
			json product = {
				{"productRID", productID },
				{"productName", productName}
			};

			jsonProducts.push_back(product);
			
			DbNext(ProductsQ );
		}
		DbCloseQuery(ProductsQ );
		DbNext(ProductsTreeQ);
	}
	std::string jStr = jsonProducts.dump();
	std::wstring widestr = from_utf8(jStr);
	return widestr;
```
and convertable functions:

```
std::wstring from_utf8(const std::string& str)
{
	std::wstring_convert<std::codecvt_utf8<wchar_t>> myconv;
	return myconv.from_bytes(str);
}

// convert wstring to UTF-8 string
std::string to_utf8(const std::wstring& str)
{
	std::wstring_convert<std::codecvt_utf8<wchar_t>> myconv;
	return myconv.to_bytes(str);
}
```

Dear developer, thanks for the unhelpful reply a week later. 
And also for the fact that you closed the issue immediately after I opened it. 
You're wrong.

**Reply by nlohmann** [reply]
I added an FAQ entry: https://json.nlohmann.me/home/faq/#wide-string-handling

**Comment by z16166** [comment]
Edit: 
I found the answer by nlohmann(or sb. else). Just add adl_serialize for std::wstring, that's it.
to_utf8/from_utf8 can be implemented by OS API、C++ std::codecvt、libiconv、ICU etc.
```
namespace nlohmann {
	
template <> struct adl_serializer<std::wstring> {
  static void to_json(json &j, const std::wstring &str) { j = to_utf8(str); }

  static void from_json(const json &j, std::wstring &str) { str = from_utf8(j.get<std::string>()); }
};

template <typename T> struct adl_serializer<std::optional<T>> {
  static void to_json(json &j, const std::optional<T> &opt) {
    if (opt == std::nullopt) {
      j = nullptr;
    } else {
      j = *opt;
    }
  }

  static void from_json(const json &j, std::optional<T> &opt) {
    if (j.is_null()) {
      opt = std::nullopt;
    } else {
      opt = j.get<T>();
    }
  }
};

}

```

How about deserializing a json string directly to a struct with std::wstring members? This is a feature request.

We store std::wstring in structs to do case-insensitive comparision/search. UTF8 string members are not good here.

struct XXX {
std::wstring a;
std::wstring b;
std::wstring c;
};

Yes, we can define a corresponding struct and manually convert UTF8 to std::wstring, but if there are so many structs, it's very inconvenient.

struct UTF8XXX {
std::string a;
std::string b;
std::string c;
};

UTF8XXX src = deserialize(json_string);

XXX dst;
dst.a = to_lower(to_wstring(src.a));
dst.b = to_lower(to_wstring(src.b));
dst.c = to_lower(to_wstring(src.c));


or, we can use std::utf16string?

**Comment by sergeniously** [comment]
@lipasite try https://github.com/sergeniously/az-json which supports std::wstring
